test_libs/pyspec/eth2spec/utils/ssz/ssz_typing.py

```diff
@@ -159,365 +119,173 @@ def get_field_types(cls):
-# -----------------------------
-
-
-def infer_type(obj):
-    if is_uint_type(obj.__class__):
-        return obj.__class__
-    elif isinstance(obj, int):
-        return uint64
-    elif isinstance(obj, list):
-        return List[infer_type(obj[0])]
-    elif isinstance(obj, (Vector, Container, bool, BytesN, bytes)):
-        return obj.__class__
-    else:
-        raise Exception("Unknown type for {}".format(obj))
-
-
-def infer_input_type(fn):
-    """
-    Decorator to run infer_type on the obj if typ argument is None
-    """
-    def infer_helper(obj, typ=None, **kwargs):
-        if typ is None:
-            typ = infer_type(obj)
-        return fn(obj, typ=typ, **kwargs)
-    return infer_helper
-
-
-def is_bool_type(typ):
-    """
-    Check if the given type is a bool.
-    """
-    if hasattr(typ, '__supertype__'):
-        typ = typ.__supertype__
-    return isinstance(typ, type) and issubclass(typ, bool)
-
-
-def is_list_type(typ):
-    """
-    Check if the given type is a list.
-    """
-    return get_origin(typ) is List or get_origin(typ) is list
+class List(AbstractList):
+    def value_check(self, value):
+        return len(value) <= self.__class__.length and super().value_check(value)
 
+    @classmethod
+    def default(cls):
+        return cls()
 
-def is_bytes_type(typ):
-    """
-    Check if the given type is a ``bytes``.
-    """
-    # Do not accept subclasses of bytes here, to avoid confusion with BytesN
-    return typ == bytes
 
+class Vector(AbstractList, metaclass=AbstractListMeta):
+    def value_check(self, value):
+        return len(value) == self.__class__.length and super().value_check(value)
 
-def is_bytesn_type(typ):
-    """
-    Check if the given type is a BytesN.
-    """
-    return isinstance(typ, type) and issubclass(typ, BytesN)
+    @classmethod
+    def default(cls):
+        return [cls.elem_type.default() for _ in range(cls.length)]
 
 
-def is_list_kind(typ):
-    """
-    Check if the given type is a kind of list. Can be bytes.
-    """
-    return is_list_type(typ) or is_bytes_type(typ)
+class BytesMeta(AbstractListMeta):
+    elem_type: DefaultingTypeMeta = byte
+    length: int
 
 
-def is_vector_type(typ):
-    """
-    Check if the given type is a vector.
-    """
-    return isinstance(typ, type) and issubclass(typ, Vector)
+class BytesLike(AbstractList, metaclass=BytesMeta):
 
+    @classmethod
+    def extract_args(cls, args):
+        if isinstance(args, bytes):
+            return args
+        elif isinstance(args, BytesLike):
+            return args.items
+        elif isinstance(args, GeneratorType):
+            return bytes(args)
+        else:
+            return bytes(args)
 
-def is_vector_kind(typ):
-    """
-    Check if the given type is a kind of vector. Can be BytesN.
-    """
-    return is_vector_type(typ) or is_bytesn_type(typ)
+    @classmethod
+    def value_check(cls, value):
+        return len(value) == cls.length and isinstance(value, bytes)
```

Comment:
Any reason why this goes here and not in `Bytes`? Seems like the latter would be more intuitive due to more symmetry? Or are we afraid of the parent class not having a `value_check` method for some reason?

Reply:
I think d83017d fixes this. BytesLike checks if it is a `bytes`, `AbstractList` goes through each element to check the type. `BytesN` and `Vector` check the length, then call the checking function of their super class.